enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration will be required for OpenStack and vSphere providers
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: No. The CCM vendors part of the code serving cloud features from the [cloud-provider](https://github.com/kubernetes/cloud-provider). CCM provides a secure TCP connection on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22) and communicates similarly to other kubernets components.
+
+Q: What happens if the kcm leader has no access to ccm? Can it continue its work? Will it give up leadership?
+A: Expect KCM perform independently to CCM and do not care about its state after initial code migration. The CCM cluster operator will be responsible for going Unavailable or Degraded in a similar situation.
+
+Q: What are the thoughts about certificate management?
+A: Certificate management should be performed by the CCM operator, like it's done in the KCM operator.
+
+Q: What happens if the KCM leader has no access to CCM? Can it continue its work? Will it give up leadership?
+A: Since KCM does not communicate with CCM it can continue to work if CCM is not available.
+
+Q: Does every node need a CCM?
+A: No. Control plane nodes only. [Source](https://kubernetes.io/docs/concepts/overview/components/#cloud-controller-manager)
+
+Q: How metrics are affected by the CCM migration?
+A: Currently the `CCM` implementation is using the same set of metrics from the core repository. The `CCMO` will require prometheus configuration to collect them from a different location.
```

Comment:
What kind of metrics are exposed, which are relevant for cluster administrator

Reply:
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/gce/metrics.go

Reply:
It was my understanding that the metrics that are being exposed haven't changed, they just need to come from a new place, since this isn't a change, do we really need to document what the particular metrics are?

Reply:
I think high-level overview of what we provide would be desirable.

Reply:
https://github.com/openshift/enhancements/pull/463/files#diff-3e0e2c48e70215076dfe36c13768a823ab7080d929d80292f37db2ef5a2121e8R231